### ISSUE TITLE
Learning: Some specs can result in infinite runtime after mutation

### ISSUE THREAD

**Issue by tjchambers** [mbj/mutant#274]
Feel free to adjust the title.

This is not an issue, but an experience I thought worth relating. 

Given:

- A class that unzips a MS Word .DOCX and builds a hash of the internal contents (name of file => content)
- The purpose of the class is to offer a set of subfiles (long xml strings) that need some manipulation, and then the .docx is later reconstructed for PDF conversion and shipping to a user

When I started running Mutant I was around 70% kills. One of the mutated methods looks like:

```diff
 def unzip
   everything = {}
   taggable = {}
   Zip::Archive.open(@path) do |ar|
     n = ar.num_files
     n.times do |i|
       name = ar.get_name(i)
       ar.fopen(name) do |f|
         content = f.read
         if (content =~ /&lt;.*tag.*id.*&gt;/)
           cleaner = DocxRepair.new.repair(content)
           unless (cleaner == content)
             ar.replace_buffer(i, cleaner)
           end
           taggable[name] = cleaner
           everything[name] = cleaner
         else
-          everything[name] = content
+          everything[name] = self
         end
       end
     end
   end
   @all_files = everything.sort.to_h
   @files = taggable.sort.to_h
 end
```

Please ignore how horrible it looks - it needs refactored. But I wanted a solid test suite first. 

Because of all of the mutations that resulted in 'malforming' the ultimate @all_files hash (changing the values in the hash, not converting it to a hash, not sorting it, not setting it to the instance variable...), on every run I was not getting anywhere with killing mutants. (This is by the way one of my learnings, that the more complex the code the harder to kill mutations - clean code = easier kills). But I was getting successful runs - runs that completed. 

Then I decided to work backward from the expected end result I so I added the following expectation to my spec:

```ruby
expect(Digest::SHA1.hexdigest(wbz.all_files.to_json)).to eq('a9ec67e1df26d83506a7a19616e04df8f775e8e6')
```

My reasoning was if I expected this outcome - a sorted hash of the files and their contents, then I could sha1 the json equivalent and compare the digest hash. Then I would ensure that the hash had the right content in the right order and would be complete if it met that requirement. That covered a lot of the bases that were giving me problems. 

EXCEPT - All of the sudden Mutant no longer completed. It would use all 8 processors for about 5 minutes and then right near the end back off to one processor, but never finished. I waited 20 minutes. Now I know that I have had code before with while loops etc. in it that when mutated never finished a run. But I never before added a spec example and caused that behavior. So I removed the spec I just added and mutant ran to completion. Added the expectation back and mutant never finished. 

Then I had an epiphany. Maybe the mutation of changing what was a string (content) -> an object (self) and then trying to take the sha1 of a hash of "self", was the issue. So I added a couple more expectations in front:

```ruby
expect(wbz.all_files).to be_a(Hash)
expect(wbz.all_files.first.last ).to be_a(String)
```

And all of the sudden mutant started running to completion again - not looping forever. I infer this was due to the precheck that failed BEFORE I got to the SHA1 of the hash of self.

So the bottom line is that some expectations can interact with the mutations in a way that the spec itself and how it is written can create an infinite run of mutant, even if the mutated code is not changed nor coded in a way the causes looping itself. 

That was my learning.

EDIT (by @mbj): Added github style markdown syntax for better code highlighting.

**Comment by mbj** [comment]
@tjchambers I read through this multiple times now. And cannot identify a real question I can write an answer for.

I'll explain how I came up with mutants mutation operators, this should help you to understand what you just learned.

The fundamental axiom behind my code tries to be: Use the least powerful primitive that does the job.

This translates to: Write no code, unless you have a reason to do so (specs).

Mutation operators I implemented for mutant help me to follow this pass. I generally mutate a complex expressions into less complex expressions. If I have no proof (specs) that the complex expression ins actually needed I'm missing a spec, or the code can be simplified.

An example is the following expression: `a; b`. Which is a sequentially executed composition of primitives `a` and primitive `b`. When mutant could remove `a` or `b` without an measurable effect on the specs one of them (or both) where redundant.

Your story is about a mutation operator that replaces an lvar read `content` with `self`. Lvars are more "complex" than a reference to `self`. Because lvars could potentially reference "any" object, where `self` is always a specific one.

This specific mutation in your control flow resulted (depending on how early your specs capture the difference) in infinite runtime.

If I write code that can potentially be mutated to infinite runtime (mutants operators are deterministic) I try to refactor my code in a way the infinite runtime cannot occur. Or I add a timeout to my test that kills such infinite runtimes. Seeing this behavior should be a warning signal for you, to not write such code (you already realized this by yourself).

I hope this explanation helps you a bit. I'm closing this issue now feel free to continue discussion here. And yes the thought I just outlined here belongs to the README, or a longer structured document.
